src/ooni/utils.cpp

```diff
@@ -76,10 +83,35 @@ ErrorOr<std::string> IPLocation::resolve_country_name(std::string ip) {
     return country_name;
 }
 
-ErrorOr<std::string> IPLocation::resolve_asn(std::string ip) {
+ErrorOr<std::string> IPLocation::resolve_city_name(std::string ip,
+                                        Var<Logger> logger) {
+    if (gi_city == nullptr) {
+        gi_city = GeoIP_open(path_city.c_str(), GEOIP_MEMORY_CACHE);
+        if (gi_country == nullptr) {
+            logger->debug("IPLocation: cannot open geoip city database");
+            return CannotOpenGeoIpCityDatabase();
+        }
+    }
+    GeoIPLookup gl;
+    GeoIPRecord *gir;
+    memset (&gl, 0, sizeof(gl));
+
+    gir = GeoIP_record_by_name(gi_city, ip.c_str());
+    if (gir == nullptr) {
+        GeoIP_delete(gi_city);
+        return GenericError();
+    }
+	std::string result = gir->city;
```

Comment:
@AntonioLangiu beware that the above line is indented using tabs rather than spaces